### ISSUE TITLE
Fixed/Persistent Session URL + Session state components

### ISSUE THREAD

**Issue by Frooxius** [Frooxius/NeosPublic#429]
With the recent need to setup bigger events due to COVID-19, I thought I'd log this idea as it's been something I wanted to add for a long while. It's relatively easy to implement and would allow for some cool new applications.

At the moment, each time a session is started it is assigned an unique session ID, which can be used to fetch info about that session and create links to it. Problem is that the ID is different every time and not predictable.

With this proposed feature, it would be possible to create/assign a fixed session ID, that the active session could always be bound to. This would make the given session always available under that ID  (and consequently URL) if it's running.

Additionally a set of components for fetching the status of that particular session would be added. This would provide information on the current state of the session - is it running? How many people are there? What's the capacity? And so on.

This would allow building hubs, with "portals" to a given session (that could be hosted by a headless). The hub could be published somewhere, and allow users to easily join in any active persistent rooms. The portal could indicate if the given session is at capacity or if it's running at all.

It would also allow basic linking of worlds together. E.g. one world could always have a portal connecting to another instance of a different world (or even the same one).

Another possibility is easier load balancing - having multiple instances and have a simple LogiX randomly pick one (or one with least amount of people) from the ones that aren't at capacity. Although there's a question if this should be handled natively instead or components could be provided to handle this since it's pretty common.

There are a few things to consider
- What form the session ID would take to avoid collision and how to prevent multiple sessions using the same one. Easiest is to just allow each user to use arbitrary name/path under their name/ID. E.g. something like Frooxius/NeosHub/Instance2.
- Related to this, we could allow registering unique "top level" session URL's, akin to domains, e.g. for a small NCR fee. But that's more for the future.
- How is it linked. I think for start just adding a config in the headless to allow to specify would help the most, as there's probably not much need to assign the session ID to random on-the-fly sessions

**Comment by TehTurk** [comment]
So would this be something akin to visiting a website? Or would it just be another version of the world orbs, or because of the persistent urls, they can be hosted, but not seen by the session

**Comment by Enverex** [comment]
Neither; it would mean that the session IDs are fixed, so that you could create world transitions in a session which would go to another session (in another world) but it would always be the same session people end up in (rather than the first person to go through hosting it as a new session). So for example if you have 3 cities, you could set up teleporters between them and people would always end up in the same sessions each time when teleporting between those sessions using those teleporters.

A more simple example would be that you could have a world orb in one of your worlds that always goes to the session of a world you're hosting on your headless, rather than it just creating a new session on the user each time. That also means that anyone else following that orb, would end up in the same session as everyone else.

Basically this change would be somewhat invisible to the normal user, it a change to how sessions are handled, in that you can create "fixed" ones rather than them being randomly generated.

I've just typed all that out and now I read it back, I'm not sure I've clarified it at all...

**Comment by Frooxius** [comment]
It could be a bit like visiting a website, where under a fixed URL there's a particular world running. It would allow for hyperlinking of live sessions too.

It wouldn't be another version of world orbs, from user perspective not much would have to be different. It would simply allow setting up links and portals to worlds that reliably go to a particular live session, without having to somehow dynamically fetch its ID and it's URL.

There's actually another feature with that - by default the worlds could be spun down and instead have a system where if the first user joins in, it sends a request to spin up that session. That could conserve resources.

**Comment by Enverex** [comment]
> There's actually another feature with that - by default the worlds could be spun down and instead have a system where if the first user joins in, it sends a request to spin up that session. That could conserve resources.

That would be incredibly useful. It would also allow you to host more worlds because you wouldn't have to worry about the passive resource usage of each idle world (which can add up quite quickly).

**Comment by Frooxius** [comment]
Finally implemented as part of 2020.7.31.1099 when refactoring the system for session announcing and related subsystems. 

- Added support for custom persistent SessionIDs in the format <userID>:<custom text> for the headless client, using new customSessionId parameter
-- This allows you to create permanent URLs and links to the session, which will always open it whenever it's running (even if you restart headless) (requested by @Rue Shejn | Artist 3D, @Nexulan, @Aegis_Wolf | Art Director, @jasonmo for Metamovie, @Enverex and many others previously)
-- This can be combined with the new "SessionInfoSource" component to update in-game objects based on the state of the session
-- userId needs to be your own (hosting user), you cannot use anyone else's
-- customText can only contain alphanumeric characters, "-" (dash) and "_" (underscore)
-- Example: U-FrooxLess:My-Custom-Session
-- Note that the actual session ID is always prefixed with "S-", for above example the session ID will be: S-U-FrooxLess:My-Custom-Session
